src/eval.rs

```diff
@@ -0,0 +1,844 @@
+        right: TreeResult,
+    },
+    BoolOr {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    InSet {
+        set: TreeResult,
+        index: TreeResult,
+    },
+    Implication {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    With {
+        value: TreeResult,
+        body: TreeResult,
+    },
+    List {
+        items: Vec<TreeResult>,
+    },
+    Apply {
+        lambda: TreeResult,
+        arg: TreeResult,
+    },
+    Map {
+        inherits: Vec<TreeResult>,
+        definitions: Vec<TreeResult>,
+    },
+    MapAttr {
+        path: Vec<TreeResult>,
+        value: TreeResult,
+    },
+    StringInterpol {
+        parts: Vec<TreeResult>,
+    },
+    UnaryInvert {
+        value: TreeResult,
+    },
+    UnaryNegate {
+        value: TreeResult,
+    },
+    Lambda {
+        literal: Gc<NixValue>,
+        params: Vec<TreeResult>,
+        body: TreeResult,
+    },
+    InferredIdent {
+        name: String,
+    },
+}
+
+#[derive(Clone, Trace, Finalize)]
+pub struct Tree {
+    #[unsafe_ignore_trace]
+    pub range: Option<TextRange>,
+    pub hash: GcCell<Option<String>>,
+    pub value: GcCell<Option<Gc<NixValue>>>,
+    pub source: TreeSource,
+    pub scope: Gc<Scope>,
+}
+impl std::fmt::Debug for Tree {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Tree")
+            .field("value", &self.value)
+            .field("source", &self.source)
+            .field("range", &self.range)
+            .finish()
+    }
+}
+
+pub fn to_hex_string(bytes: Vec<u8>) -> String {
+    let strs: Vec<String> = bytes.iter().map(|b| format!("{:02x}", b)).collect();
+    strs.join("")
+}
+
+impl Tree {
+    pub fn hash(&self) -> Result<String, EvalError> {
+        use std::ops::Deref;
+        let value_borrow = self.hash.borrow();
+        if let Some(ref value) = value_borrow.deref() {
+            Ok(value.clone())
+        } else {
+            drop(value_borrow);
+            let value = self.hash_uncached()?;
+            *self.hash.borrow_mut() = Some(value.clone());
+            Ok(value)
+        }
+    }
+
+    pub fn completions(&self) -> Option<(String, Vec<String>, TextRange)> {
+        match &self.source {
+            TreeSource::Ident { name } => Some((name.clone(), self.scope.list(), self.range?)),
+            TreeSource::Select { from, index } => Some((
+                index.as_ref().ok()?.get_ident().ok()?,
+                if let Ok(x) = from {
+                    x.get_keys().unwrap_or_default()
+                } else {
+                    vec![]
+                },
+                index.as_ref().ok()?.range?,
+            )),
+            _ => Some(("".to_string(), self.scope.list(), self.range?)),
+        }
+    }
+
+    pub fn get_keys(&self) -> Result<Vec<String>, EvalError> {
+        Ok(self.eval()?.as_map()?.keys().cloned().collect())
+    }
+
+    fn hash_uncached(&self) -> Result<String, EvalError> {
```

Comment:
Hashing (explained in the PR description) is an ugly hack at the moment. This PR is mostly just to show how integration with an evaluator would work if rnix-lsp wanted to do so.